Source/WebCore/Modules/webcodecs/WebCodecsAudioData.cpp

```diff
@@ -0,0 +1,187 @@
+#include "WebCodecsAudioDataAlgorithms.h"
+
+namespace WebCore {
+
+// https://www.w3.org/TR/webcodecs/#dom-audiodata-audiodata
+ExceptionOr<Ref<WebCodecsAudioData>> WebCodecsAudioData::create(ScriptExecutionContext& context, Init&& init)
+{
+    if (!isValidAudioDataInit(init))
+        return Exception { TypeError, "Invalid init data"_s };
+
+    auto rawData = init.data.span();
+    auto data = PlatformRawAudioData::create(WTFMove(rawData), init.format, init.sampleRate, init.timestamp, init.numberOfFrames, init.numberOfChannels);
+
+    return adoptRef(*new WebCodecsAudioData(context, WebCodecsAudioInternalData { WTFMove(data) }));
+}
+
+Ref<WebCodecsAudioData> WebCodecsAudioData::create(ScriptExecutionContext& context, Ref<PlatformRawAudioData>&& data)
+{
+    return adoptRef(*new WebCodecsAudioData(context, WebCodecsAudioInternalData { WTFMove(data) }));
+}
+
+WebCodecsAudioData::WebCodecsAudioData(ScriptExecutionContext& context)
+    : ContextDestructionObserver(&context)
+{
+}
+
+WebCodecsAudioData::WebCodecsAudioData(ScriptExecutionContext& context, WebCodecsAudioInternalData&& data)
+    : ContextDestructionObserver(&context)
+    , m_data(WTFMove(data))
+{
+}
+
+WebCodecsAudioData::~WebCodecsAudioData()
+{
+    if (m_isDetached)
+        return;
+    if (auto* context = scriptExecutionContext()) {
+        context->postTask([](auto& context) {
+            context.addConsoleMessage(MessageSource::JS, MessageLevel::Warning, "A AudioData was destroyed without having been closed explicitly"_s);
+        });
+    }
+}
+
+std::optional<AudioSampleFormat> WebCodecsAudioData::format() const
+{
+    if (!m_data.audioData)
+        return { };
+
+    return m_data.audioData->format();
+}
+
+float WebCodecsAudioData::sampleRate() const
+{
+    if (!m_data.audioData)
+        return 0;
+
+    return m_data.audioData->sampleRate();
+}
+
+size_t WebCodecsAudioData::numberOfFrames() const
+{
+    if (!m_data.audioData)
+        return 0;
+
+    return m_data.audioData->numberOfFrames();
+}
+
+size_t WebCodecsAudioData::numberOfChannels() const
+{
+    if (!m_data.audioData)
+        return 0;
+
+    return m_data.audioData->numberOfChannels();
+}
+
+std::optional<uint64_t> WebCodecsAudioData::duration()
+{
+    if (!m_data.audioData)
+        return { };
+
+    return m_data.audioData->duration();
+}
+
+int64_t WebCodecsAudioData::timestamp() const
+{
+    if (!m_data.audioData)
+        return 0;
+
+    return m_data.audioData->timestamp();
+}
+
+// https://www.w3.org/TR/webcodecs/#dom-audiodata-allocationsize
+ExceptionOr<size_t> WebCodecsAudioData::allocationSize(const CopyToOptions& options)
+{
+    if (isDetached())
+        return Exception { InvalidStateError, "AudioData is detached"_s };
+
+    auto copyElementCount = computeCopyElementCount(*this, options);
+    if (copyElementCount.hasException())
+        return copyElementCount.releaseException();
+
+    auto destFormat = options.format.value_or(*format());
```

Comment:
In theory, there is nothing precluding having m_isDetached=false by m_data.audioData be null.

Reply:
I'm sorry I don't understand this comment :)

Reply:
How is it guaranteed that format() is not std::nullopt here?

Reply:
Ah right, I understand now, thanks

Reply:
I refactored `computeCopyElementCount()` to raise an exception earlier if internal data is null, so I think we can keep this as-is.

Reply:
Maybe computeCopyElementCount could also return the destination format as well.
computeCopyElementCount uses method that repeatedly checks internal data is not null, but I guess it is fine as is.